pmlc/dialect/stripe/autostencil_pass.cc

```diff
@@ -0,0 +1,312 @@
+  }
+  IVLOG(3, "Middle: loop = " << tot_middle_loop);
+  for (auto& kvp : middle_idxs) {
+    if (kvp.second > 1) {
+      IVLOG(3, idxName(kvp.first).str() << ": " << kvp.second);
+    }
+  }
+
+  llvm::DenseMap<BlockArgument*, size_t> outer_idxs; 
+  for (auto idx : out_idxs) {
+    outer_idxs.try_emplace(idx, idxRange(idx));
+  }
+  for (size_t i = 0; i < numIndex; ++i) {
+    auto it = outer_idxs.find(inner_idxs[i]);
+    if (it != outer_idxs.end()) {
+      it->second = (it->second - 1) / tiles[i] + 1;
+    }
+  }
+  size_t tot_outer_loop = 1;
+  for (auto& kvp : outer_idxs) {
+    tot_outer_loop *= kvp.second;
+  }
+  IVLOG(3, "Outer: loop = " << tot_outer_loop);
+  for (auto& kvp : outer_idxs) {
+    if (kvp.second > 1) {
+      IVLOG(3, idxName(kvp.first).str() << ": " << kvp.second);
+    }
+  }
+
+  size_t outer_batches = (tot_outer_loop - 1) / n_cpus + 1;
+  double perf = outer_batches * tot_middle_loop * inner_time;
+  IVLOG(3, "Performance = " << perf);
+
+  return perf;
+}
+
+void AutoStencil::SearchTiles(size_t idx) {
+  if (idx >= numIndex) {
+    double performance = Evaluate();
+    if (performance > best_perf) {
+      best_perf = performance;
+      best_tiles.clear();
+      for (size_t i = 0; i < numIndex; ++i) {
+        best_tiles[idxName(inner_idxs[i])] = tiles[i];
+      }
+    }
+    return;
+  }
+  size_t range = idxRange(inner_idxs[idx]);
+  for (size_t i = 2; i <= range; ++i) {
+    // TODO: consider po2
+    if (range % i == 0) {
+      tiles[idx] = i;
+      SearchTiles(idx + 1);
+    }
+  }
+}
+
+bool AutoStencil::ConflictInnerIndex(BlockArgument* idx) {
+  BlockArgumentSet& conflict_set = conflict[idx];
+  for (auto& elem : inner_idxs) {
+    if (elem == idx) {
+      return true;
+    }
+    if (conflict_set.find(elem) != conflict_set.end()) {
+      return true;
+    }
+  }
+  return false;
+}
+
+// Search for matrix index (0 for M, 1 for N, 2 for K)
+void AutoStencil::SearchIndex(size_t matrix_idx) {
+  if (matrix_idx >= numIndex) {
+    SearchTiles(0);
+    return;
+  }
+  auto& idxs = (matrix_idx == 2) ? all_idxs : out_idxs;
+  for (auto idx : idxs) {
+    if (!ConflictInnerIndex(idx)) {
+      inner_idxs.push_back(idx);
+      SearchIndex(matrix_idx + 1);
+      inner_idxs.pop_back();
+    }
+  }
+}
+
+// Collect the index used by value's RefineOp
+// If with_conflict is true, specify the conflict between index, i.e., both index can't be in the inner block
+BlockArgumentSet AutoStencil::RefUsedIdxs(Value* value, bool with_conflict) {
+  BlockArgumentSet used_idxs;
+  auto ref_op = mlir::dyn_cast<RefineOp>(value->getDefiningOp());
+  auto tensor = value->getType().cast<TensorRefType>();
+  if (!ref_op) {
+    throw std::runtime_error("Access a tensor not defined by RefineOp.");
+  }
+  for (int i = 0; i < tensor.getRank(); i++) {
+    auto access = AffinePolynomial(ref_op.getOffset(i));
+    for (auto [arg0, scale0] : access.terms) {
+      used_idxs.insert(arg0);
+      if (with_conflict) {
```

Comment:
If I understand correctly, the conflict sets prevent one from picking say both 'i' and 'x' for a convolution which has D[i + x] somewhere?  So basically any index conflicts with any other index that appears in the same polynomial at any point, right?  That makes a lot of sense, but some comments about that when the conflict member variables are defined would be useful, since it took me a while to understand.